sklearn/linear_model/_linear_loss.py

```diff
@@ -0,0 +1,362 @@
+            Training data.
+        sample_weight : None or C/F-contiguous array of shape (n_samples,)
+            Sample weights.
+        alpha: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+        """
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss = self._loss.loss(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=n_threads,
+        )
+        loss = loss.sum()
+
+        if w.ndim == 1:
+            return loss + 0.5 * alpha * (w @ w)
+        else:
+            return loss + 0.5 * alpha * squared_norm(w)
+
+    def loss_gradient(self, coef, X, y, sample_weight=None, alpha=0.0, n_threads=1):
+        """Computes the sum/average of loss and gradient.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+        y : C/F-contiguous array of shape (n_samples,)
+            Observed, true target values.
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+        sample_weight : None or C/F-contiguous array of shape (n_samples,)
+            Sample weights.
+        alpha: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+
+        gradient : ndarray of shape (n_dof,) or (n_classes * n_dof)
+             The gradient of the loss as ravelled array.
+        """
+        n_features, n_classes = X.shape[1], self._loss.n_classes
+        n_dof = n_features + self.fit_intercept
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss, gradient = self._loss.loss_gradient(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=n_threads,
+        )
+        loss = loss.sum()
+
+        if not self._loss.is_multiclass:
+            loss += 0.5 * alpha * (w @ w)
+            grad = np.empty_like(coef, dtype=X.dtype)
+            grad[:n_features] = X.T @ gradient + alpha * w
+            if self.fit_intercept:
+                grad[-1] = gradient.sum()
+            return loss, grad
+        else:
+            loss += 0.5 * alpha * squared_norm(w)
+            grad = np.empty((n_classes, n_dof), dtype=X.dtype)
+            # gradient.shape = (n_samples, n_classes)
+            grad[:, :n_features] = gradient.T @ X + alpha * w
+            if self.fit_intercept:
+                grad[:, -1] = gradient.sum(axis=0)
+            return loss, grad.ravel()
```

Comment:
```python
%%magic
```
This and similar code segments is where the *magic* happens.
-  The loss itself just needs the linead predictor `raw_prediction=X @ coef`.
- Gradients and hessians need the inner derivative of the linear predictor w.r.t. the coefficients, i.e. **multiply the output of `loss.gradient` by `X`**.
- L2 penalty is added.

Reply:
Would you change it now or wait until monitoring callbacks arrive? This class will become a central place, so a change would be more or less easy.

Reply:
Fine with waiting for later.